File: LedgerTargetMCP/src/propertyManagementReport.ts
import { createHash } from "node:crypto";
import { TargetMCPFailure, validateIdentifier } from "./contractSupport.js";

export type PropertyManagementReportProject = Readonly<{
  accountId: string; projectId: string; name: string; address: string | null; revision: string;
}>;
export type PropertyManagementReportSpace = Readonly<{
  accountId: string; projectId: string; spaceId: string; name: string; revision: string;
}>;
export type PropertyManagementReportItem = Readonly<{
  accountId: string; projectId: string; itemId: string; placementId: string;
  spaceId: string | null; name: string; sku: string | null; itemRevision: string;
  marketValueMinorUnits: string | null; marketValueCurrency: string | null;
}>;
export type ProjectItemAccountingRow = Readonly<{
  evidence: Readonly<{
    accountId: string; projectId: string; clientId: string; itemId: string; spaceId?: string | null;
    clientPaidPurchases: readonly Readonly<{
      id: string; accountId: string; projectId: string; clientId: string; itemId: string; transactionId: string;
      classification: Readonly<{ type: "purchase"; role: "standalone";
        scope: Readonly<{ ownerKind: "project"; accountId: string; projectId: string; clientId: string }> }>;
    }>[];
    billableOccurrences: readonly Readonly<{
      id: string; accountId: string; projectId: string; itemId: string; polarity: "charge" | "credit";
      phase: Readonly<{ kind: "availableToInvoice" | "onLiveInvoice" | "frozenPaid"; invoiceId?: string | null }>;
    }>[];
  }>;
  relationshipAbsenceIsAuthoritative: boolean;
  resolution: "accountedFor" | "unaccountedFor" | "relationshipEvidenceIncomplete";
}>;
export type PropertyManagementReportItemInput = PropertyManagementReportItem & Readonly<{
  accounting?: ProjectItemAccountingRow | null;
}>;
export type AuthoritativePropertyManagementReportProvenance = Readonly<{
  accountId: string; projectId: string; principalId: string; visibilityScopeID: string;
  source: Readonly<{ kind: "authoritative" }>;
  authorityVersion: string; asOf: number; readiness: "ready";
}>;
export type PropertyManagementReportInput = Readonly<{
  project: PropertyManagementReportProject;
  spaces: readonly PropertyManagementReportSpace[];
  items: readonly PropertyManagementReportItemInput[];
  currency: string;
  provenance: AuthoritativePropertyManagementReportProvenance;
}>;
export type PropertyManagementReportTotals = Readonly<{
  itemCount: number; knownMarketValueSubtotalMinorUnits: string; currency: string;
  unknownMarketValueCount: number; totalMarketValueMinorUnits: string | null;
}>;
export type PropertyManagementReportGroup = Readonly<{
  // Swift's synthesized Optional encoding omits this key for No Space.
  spaceId?: string; name: string; rows: readonly PropertyManagementReportItem[];
  totals: PropertyManagementReportTotals;
}>;
export type PropertyManagementReportContent = Readonly<{
  reportKind: "property_management"; project: PropertyManagementReportProject;
  provenance: AuthoritativePropertyManagementReportProvenance; currency: string;
  spaces: readonly PropertyManagementReportSpace[]; groups: readonly PropertyManagementReportGroup[];
  totals: PropertyManagementReportTotals; sourceSetHash: string;
}>;
export type PropertyManagementReportSnapshot = PropertyManagementReportContent & Readonly<{
  reference: Readonly<{
    snapshotID: string; snapshotHash: string; visibilityScopeID: string;
    profileVersion: "property-management-v1"; authorityVersion: string;
  }>;
}>;

const MIN = -(1n << 63n), MAX = (1n << 63n) - 1n, UINT_MAX = (1n << 64n) - 1n;
function fail(code: string): never { throw new TargetMCPFailure(`property_report_${code}`); }
function string(value: unknown): string {
  if (typeof value !== "string") fail("invalid_text");
  // Foundation strings cannot contain unpaired UTF-16 surrogates.
  for (let i = 0; i < value.length; i++) {
    const code = value.charCodeAt(i);
    if (code >= 0xd800 && code <= 0xdbff) {
      const low = value.charCodeAt(++i);
      if (!(low >= 0xdc00 && low <= 0xdfff)) fail("invalid_text");
    } else if (code >= 0xdc00 && code <= 0xdfff) fail("invalid_text");
  }
  return value;
}
function nullableText(value: unknown): string | null { return value === null ? null : string(value); }
function id(value: unknown): string { return validateIdentifier(string(value), "property_report_invalid_identifier"); }
function revision(value: unknown): string {
  if (typeof value !== "string" || !/^[1-9][0-9]*$/.test(value) || value.length > 20 || BigInt(value) > UINT_MAX) fail("invalid_revision");
  return value;
}
function amount(value: unknown): string {
  if (typeof value !== "string" || !/^(0|-[1-9][0-9]*|[1-9][0-9]*)$/.test(value) || value.length > 20) fail("invalid_amount");
  const number = BigInt(value);
  if (number < MIN || number > MAX) fail("invalid_amount");
  return value;
}
function currency(value: unknown): string {
  if (typeof value !== "string" || !/^[A-Z]{3}$/.test(value)) fail("invalid_currency");
  return value;
}
function scope(accountId: unknown, projectId: unknown, project: PropertyManagementReportProject): void {
  if (id(accountId) !== project.accountId || id(projectId) !== project.projectId) fail("scope_mismatch");
}
function unique(set: Set<string>, value: string, code: string): void {
  if (set.has(value)) fail(code);
  set.add(value);
}
function ordered(nameA: string, idA: string, nameB: string, idB: string): number {
  return Buffer.compare(Buffer.from(nameA), Buffer.from(nameB)) || Buffer.compare(Buffer.from(idA), Buffer.from(idB));
}
function record(value: unknown): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) fail("invalid_accounting");
  return value as Record<string, unknown>;
}
/** Decode the same constrained relationships as Swift, rebuilding derived state
 * and optional fields before hashing; a caller-supplied resolution is not proof. */
function accounting(value: unknown, item: PropertyManagementReportItem): ProjectItemAccountingRow {
  if (value === undefined || value === null) fail("incomplete_readiness");
  const row = record(value), evidence = record(row.evidence);
  const accountId = id(evidence.accountId), projectId = id(evidence.projectId);
  const clientId = id(evidence.clientId), itemId = id(evidence.itemId);
  const spaceId = evidence.spaceId == null ? null : id(evidence.spaceId);
  if (!Array.isArray(evidence.clientPaidPurchases) || !Array.isArray(evidence.billableOccurrences)
      || typeof row.relationshipAbsenceIsAuthoritative !== "boolean") fail("invalid_accounting");
  function relationshipScope(connection: Record<string, unknown>): void {
    if (id(connection.accountId) !== accountId || id(connection.projectId) !== projectId
        || id(connection.itemId) !== itemId) fail("scope_mismatch");
  }
  const purchaseIDs = new Set<string>(), occurrenceIDs = new Set<string>();
  const clientPaidPurchases = evidence.clientPaidPurchases.map(raw => {
    const connection = record(raw), classification = record(connection.classification);
    const purchaseScope = record(classification.scope);
    relationshipScope(connection);
    if (id(connection.clientId) !== clientId || id(purchaseScope.accountId) !== accountId
        || id(purchaseScope.projectId) !== projectId || id(purchaseScope.clientId) !== clientId) fail("scope_mismatch");
    if (classification.type !== "purchase" || classification.role !== "standalone"
        || purchaseScope.ownerKind !== "project") fail("invalid_accounting");
    const connectionId = id(connection.id);
    unique(purchaseIDs, connectionId, "duplicate_accounting_relationship");
    return { id: connectionId, accountId, projectId, clientId, itemId, transactionId: id(connection.transactionId),
      classification: { type: "purchase" as const, role: "standalone" as const,
        scope: { ownerKind: "project" as const, accountId, projectId, clientId } } };
  });
  const billableOccurrences = evidence.billableOccurrences.map((raw): ProjectItemAccountingRow["evidence"]["billableOccurrences"][number] => {
    const occurrence = record(raw), phase = record(occurrence.phase);
    relationshipScope(occurrence);
    const occurrenceId = id(occurrence.id);
    unique(occurrenceIDs, occurrenceId, "duplicate_accounting_relationship");
    const polarity = occurrence.polarity;
    if (polarity !== "charge" && polarity !== "credit") fail("invalid_accounting");
    const kind = phase.kind, invoiceId = phase.invoiceId == null ? null : id(phase.invoiceId);
    if (kind !== "availableToInvoice" && kind !== "onLiveInvoice" && kind !== "frozenPaid") fail("invalid_accounting");
    if ((kind === "availableToInvoice") !== (invoiceId === null)) fail("invalid_accounting");
    return { id: occurrenceId, accountId, projectId, itemId, polarity,
      phase: { kind, ...(invoiceId === null ? {} : { invoiceId }) } };
  });
  const resolution = clientPaidPurchases.length || billableOccurrences.length ? "accountedFor"
    : row.relationshipAbsenceIsAuthoritative ? "unaccountedFor" : "relationshipEvidenceIncomplete";
  if (row.resolution !== resolution) fail("invalid_accounting");
  if (resolution === "relationshipEvidenceIncomplete") fail("incomplete_readiness");
  if (accountId !== item.accountId || projectId !== item.projectId || itemId !== item.itemId
      || spaceId !== item.spaceId) fail("scope_mismatch");
  return { evidence: { accountId, projectId, clientId, itemId, ...(spaceId === null ? {} : { spaceId }),
    clientPaidPurchases, billableOccurrences }, relationshipAbsenceIsAuthoritative: row.relationshipAbsenceIsAuthoritative, resolution };
}
function summarize(rows: readonly PropertyManagementReportItem[], currency: string): PropertyManagementReportTotals {
  let total = 0n, unknown = 0;
  for (const row of rows) {
    if (row.marketValueMinorUnits === null) unknown++;
    else {
      total += BigInt(row.marketValueMinorUnits);
      // Match Swift checked addition at each step, not just the final sum.
      if (total < MIN || total > MAX) fail("arithmetic_overflow");
    }
  }
  return { itemCount: rows.length, knownMarketValueSubtotalMinorUnits: total.toString(), currency,
    unknownMarketValueCount: unknown, totalMarketValueMinorUnits: unknown === 0 ? total.toString() : null };
}

/** Pure domain projection. Calling this does not authenticate facts or grant
 * access. The online provider must establish membership and a coherent read. */
export function buildPropertyManagementReportSnapshot(input: PropertyManagementReportInput): PropertyManagementReportSnapshot {
  const project: PropertyManagementReportProject = {
    accountId: id(input.project.accountId), projectId: id(input.project.projectId),
    name: string(input.project.name), address: nullableText(input.project.address), revision: revision(input.project.revision),
  };
  const code = currency(input.currency), source = input.provenance;
  scope(source.accountId, source.projectId, project);
  if (source.readiness !== "ready") fail("incomplete_readiness");
  if (source.source?.kind !== "authoritative" || Object.keys(source.source).some(key => key !== "kind")
      || "lastSyncedAt" in source || "localDataVersion" in source) fail("invalid_source");
  if (typeof source.visibilityScopeID !== "string" || !/^[a-f0-9]{64}$/.test(source.visibilityScopeID)) fail("invalid_visibility_scope");
  if (typeof source.authorityVersion !== "string" || !/^[a-z0-9_.-]{1,64}$/.test(source.authorityVersion)) fail("invalid_authority_version");
  if (!Number.isSafeInteger(source.asOf) || source.asOf <= 0) fail("invalid_timestamp");
  const provenance: AuthoritativePropertyManagementReportProvenance = {
    accountId: project.accountId, projectId: project.projectId, principalId: id(source.principalId),
    visibilityScopeID: source.visibilityScopeID, source: { kind: "authoritative" },
    authorityVersion: source.authorityVersion, asOf: source.asOf, readiness: "ready",
  };
  if (!Array.isArray(input.spaces) || !Array.isArray(input.items)) fail("invalid_rows");
  const spaceIDs = new Set<string>(), itemIDs = new Set<string>(), placementIDs = new Set<string>();
  const spaces: PropertyManagementReportSpace[] = input.spaces.map(row => {
    scope(row.accountId, row.projectId, project);
    const spaceId = id(row.spaceId);
    unique(spaceIDs, spaceId, "duplicate_space");
    return { accountId: project.accountId, projectId: project.projectId, spaceId,
      name: string(row.name), revision: revision(row.revision) };
  }).sort((a, b) => ordered(a.name, a.spaceId, b.name, b.spaceId));
  const accountingRows = new Map<string, ProjectItemAccountingRow>();
  const items: PropertyManagementReportItem[] = input.items.map(row => {
    scope(row.accountId, row.projectId, project);
    const itemId = id(row.itemId), placementId = id(row.placementId);
    unique(itemIDs, itemId, "duplicate_item"); unique(placementIDs, placementId, "duplicate_placement");
    const spaceId = row.spaceId === null ? null : id(row.spaceId);
    if (spaceId !== null && !spaceIDs.has(spaceId)) fail("missing_space");
    const value = row.marketValueMinorUnits === null ? null : amount(row.marketValueMinorUnits);
    const valueCurrency = row.marketValueCurrency === null ? null : currency(row.marketValueCurrency);
    if ((value === null) !== (valueCurrency === null)) fail("invalid_amount");
    const item = { accountId: project.accountId, projectId: project.projectId, itemId, placementId, spaceId,
      name: string(row.name), sku: nullableText(row.sku), itemRevision: revision(row.itemRevision),
      marketValueMinorUnits: value, marketValueCurrency: valueCurrency };
    const eligibility = accounting(row.accounting, item);
    if (eligibility.resolution === "accountedFor" && valueCurrency !== null && valueCurrency !== code) fail("mixed_currency");
    accountingRows.set(itemId, eligibility);
    return item;
  }).filter(item => accountingRows.get(item.itemId)!.resolution === "accountedFor")
    .sort((a, b) => ordered(a.name, a.itemId, b.name, b.itemId));
  const groups: PropertyManagementReportGroup[] = [];
  for (const space of spaces) {
    const rows = items.filter(item => item.spaceId === space.spaceId);
    if (rows.length) groups.push({ spaceId: space.spaceId, name: space.name, rows, totals: summarize(rows, code) });
  }
  const unplaced = items.filter(item => item.spaceId === null);
  if (unplaced.length) groups.push({ name: "No Space", rows: unplaced, totals: summarize(unplaced, code) });
  const content: PropertyManagementReportContent = {
    reportKind: "property_management", project, provenance, currency: code, spaces, groups,
    totals: summarize(items, code), sourceSetHash: hash(canonical({ project, spaces, items,
      accountingEvidenceHash: hash(canonical([...accountingRows.entries()]
        .sort(([a], [b]) => ordered("", a, "", b)).map(([, row]) => row))) })),
  };
  const snapshotHash = hash(canonical(content));
  return { ...content, reference: { snapshotID: snapshotHash.slice(0, 32), snapshotHash,
    visibilityScopeID: provenance.visibilityScopeID, profileVersion: "property-management-v1",
    authorityVersion: provenance.authorityVersion } };
}

export function encodePropertyManagementReportSnapshot(snapshot: PropertyManagementReportSnapshot): string {
  return canonical(snapshot);
}
export function encodePropertyManagementReportContent(snapshot: PropertyManagementReportSnapshot): string {
  const { reference: _, ...content } = snapshot;
  return canonical(content);
}
function hash(bytes: string): string { return createHash("sha256").update(bytes, "utf8").digest("hex"); }
// All schema keys are ASCII. Match JSONEncoder.sortedKeys + withoutEscapingSlashes;
// existing contractSupport.canonicalJSON deliberately uses different slash rules.
function canonical(value: unknown): string {
  if (value === null || typeof value === "string" || typeof value === "boolean") return JSON.stringify(value);
  if (typeof value === "number" && Number.isSafeInteger(value)) return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (typeof value === "object" && value !== null) {
    const record = value as Record<string, unknown>;
    return `{${Object.keys(record).sort().map(key => `${JSON.stringify(key)}:${canonical(record[key])}`).join(",")}}`;
  }
  fail("encoding_invalid");
}

File: LedgeriOS/LedgerTargetCore/PropertyManagementReport.swift
import Foundation

/// Downloaded app read boundary. Online MCP uses the same snapshot content but
/// authoritative source evidence, never fabricated download readiness.
public protocol PropertyManagementReportReading: Sendable {
    func readDownloadedPropertyManagementReport(accountId: AccountID, projectId: ProjectID,
        currency: CurrencyCode, asOf: ProtectedArtifactEpochMilliseconds) async throws -> PropertyManagementReportSnapshot
}

public enum PropertyManagementReportUpdate: Equatable, Sendable {
    case incomplete
    case ready(PropertyManagementReportSnapshot)
}

public protocol PropertyManagementReportWatching: Sendable {
    func watchPropertyManagementReport(accountId: AccountID, projectId: ProjectID,
        currency: CurrencyCode) -> AsyncThrowingStream<PropertyManagementReportUpdate, Error>
}

public enum PropertyManagementReportFailure: Error, Equatable, Sendable {
    case incompleteReadiness, scopeMismatch, invalidRevision
    case duplicateItem, duplicatePlacement, duplicateSpace, missingSpace, mixedCurrency
}

public struct PropertyManagementReportProject: Encodable, Equatable, Sendable {
    public let accountId: AccountID
    public let projectId: ProjectID
    public let name: String
    public let address: String?
    public let revision: UInt64

    public init(accountId: AccountID, projectId: ProjectID, name: String, address: String?, revision: UInt64) {
        self.accountId = accountId; self.projectId = projectId
        self.name = name; self.address = address; self.revision = revision
    }
    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(accountId, forKey: .accountId)
        try values.encode(projectId, forKey: .projectId)
        try values.encode(name, forKey: .name)
        try values.encode(address, forKey: .address)
        try values.encode(String(revision), forKey: .revision)
    }
    private enum CodingKeys: String, CodingKey { case accountId, projectId, name, address, revision }
}

public struct PropertyManagementReportSpace: Encodable, Equatable, Sendable {
    public let accountId: AccountID
    public let projectId: ProjectID
    public let spaceId: SpaceID
    public let name: String
    public let revision: UInt64

    public init(accountId: AccountID, projectId: ProjectID, spaceId: SpaceID, name: String, revision: UInt64) {
        self.accountId = accountId; self.projectId = projectId; self.spaceId = spaceId
        self.name = name; self.revision = revision
    }
    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(accountId, forKey: .accountId)
        try values.encode(projectId, forKey: .projectId)
        try values.encode(spaceId, forKey: .spaceId)
        try values.encode(name, forKey: .name)
        try values.encode(String(revision), forKey: .revision)
    }
    private enum CodingKeys: String, CodingKey { case accountId, projectId, spaceId, name, revision }
}

public struct PropertyManagementReportItem: Encodable, Equatable, Sendable {
    public let accountId: AccountID
    public let projectId: ProjectID
    public let itemId: ItemID
    public let placementId: EntityID
    public let spaceId: SpaceID?
    public let name: String
    public let sku: String?
    public let marketValue: Money?
    public let itemRevision: UInt64
    public let accounting: ProjectItemAccountingRow?

    public init(accountId: AccountID, projectId: ProjectID, itemId: ItemID, placementId: EntityID,
                spaceId: SpaceID?, name: String, sku: String?, marketValue: Money?, itemRevision: UInt64,
                accounting: ProjectItemAccountingRow? = nil) {
        self.accountId = accountId; self.projectId = projectId; self.itemId = itemId; self.placementId = placementId
        self.spaceId = spaceId; self.name = name; self.sku = sku; self.marketValue = marketValue; self.itemRevision = itemRevision
        self.accounting = accounting
    }

    // JSON consumers must not round exact cents through a JavaScript Number.
    // Null stays explicit; a zero-valued Money is a known value.
    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(accountId, forKey: .accountId)
        try values.encode(projectId, forKey: .projectId)
        try values.encode(itemId, forKey: .itemId)
        try values.encode(placementId, forKey: .placementId)
        try values.encode(spaceId, forKey: .spaceId)
        try values.encode(name, forKey: .name)
        try values.encode(sku, forKey: .sku)
        try values.encode(marketValue.map { String($0.minorUnits) }, forKey: .marketValueMinorUnits)
        try values.encode(marketValue?.currency, forKey: .marketValueCurrency)
        try values.encode(String(itemRevision), forKey: .itemRevision)
    }
    private enum CodingKeys: String, CodingKey {
        case accountId, projectId, itemId, placementId, spaceId, name, sku
        case marketValueMinorUnits, marketValueCurrency, itemRevision
    }
}

public enum PropertyManagementReportSource: Encodable, Equatable, Sendable {
    case downloaded(localDataVersion: LocalDataVersion, lastSyncedAt: ProtectedArtifactEpochMilliseconds)
    case authoritative

    public var kind: String {
        switch self { case .downloaded: "downloaded"; case .authoritative: "authoritative" }
    }

    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(kind, forKey: .kind)
        if case .downloaded(let version, let syncedAt) = self {
            try values.encode(version, forKey: .localDataVersion)
            try values.encode(syncedAt, forKey: .lastSyncedAt)
        }
    }
    private enum CodingKeys: String, CodingKey { case kind, localDataVersion, lastSyncedAt }
}

/// Provider evidence, not an authorization grant. Both sources require current
/// scope authorization and one coherent SQL read of Items/placements/parents.
/// Downloaded data additionally requires the exact retained stream checkpoint.
/// Public construction and `.ready` do not establish those provider guarantees.
public struct PropertyManagementReportProvenance: Encodable, Equatable, Sendable {
    public let accountId: AccountID
    public let projectId: ProjectID
    public let principalId: PrincipalID
    public let visibilityScopeID: ProtectedArtifactVisibilityScopeID
    public let source: PropertyManagementReportSource
    public let authorityVersion: ProtectedArtifactAuthorityVersion
    public let asOf: ProtectedArtifactEpochMilliseconds
    public let readiness: ListSnapshotQuality
    public var localDataVersion: LocalDataVersion? {
        if case .downloaded(let version, _) = source { return version }
        return nil
    }
    public var lastSyncedAt: ProtectedArtifactEpochMilliseconds? {
        if case .downloaded(_, let timestamp) = source { return timestamp }
        return nil
    }

    public init(accountId: AccountID, projectId: ProjectID, principalId: PrincipalID,
                visibilityScopeID: ProtectedArtifactVisibilityScopeID, localDataVersion: LocalDataVersion,
                authorityVersion: ProtectedArtifactAuthorityVersion, asOf: ProtectedArtifactEpochMilliseconds,
                readiness: ListSnapshotQuality, lastSyncedAt: ProtectedArtifactEpochMilliseconds) {
        self.init(accountId: accountId, projectId: projectId, principalId: principalId,
            visibilityScopeID: visibilityScopeID, source: .downloaded(localDataVersion: localDataVersion, lastSyncedAt: lastSyncedAt),
            authorityVersion: authorityVersion, asOf: asOf, readiness: readiness)
    }

    public init(accountId: AccountID, projectId: ProjectID, principalId: PrincipalID,
                visibilityScopeID: ProtectedArtifactVisibilityScopeID, source: PropertyManagementReportSource,
                authorityVersion: ProtectedArtifactAuthorityVersion, asOf: ProtectedArtifactEpochMilliseconds,
                readiness: ListSnapshotQuality) {
        self.accountId = accountId; self.projectId = projectId; self.principalId = principalId
        self.visibilityScopeID = visibilityScopeID; self.source = source
        self.authorityVersion = authorityVersion; self.asOf = asOf
        self.readiness = readiness
    }
}

public struct PropertyManagementReportTotals: Encodable, Equatable, Sendable {
    public let itemCount: Int
    public let knownMarketValueSubtotal: Money
    public let unknownMarketValueCount: Int
    /// A partial known subtotal is never labelled as a complete valuation.
    public var totalMarketValue: Money? { unknownMarketValueCount == 0 ? knownMarketValueSubtotal : nil }

    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(itemCount, forKey: .itemCount)
        try values.encode(String(knownMarketValueSubtotal.minorUnits), forKey: .knownMarketValueSubtotalMinorUnits)
        try values.encode(knownMarketValueSubtotal.currency, forKey: .currency)
        try values.encode(unknownMarketValueCount, forKey: .unknownMarketValueCount)
        try values.encode(totalMarketValue.map { String($0.minorUnits) }, forKey: .totalMarketValueMinorUnits)
    }
    private enum CodingKeys: String, CodingKey {
        case itemCount, knownMarketValueSubtotalMinorUnits, currency, unknownMarketValueCount, totalMarketValueMinorUnits
    }
}

public struct PropertyManagementReportGroup: Encodable, Equatable, Sendable {
    /// Nil identifies No Space; an actual Space named "No Space" remains distinct.
    public let spaceId: SpaceID?
    public let name: String
    public let rows: [PropertyManagementReportItem]
    public let totals: PropertyManagementReportTotals
}

/// One concrete projection for preview, PDF, print, CSV and MCP. Its factory
/// checks internal coherence only. It does not authorize exports, authenticate
/// checkpoints or promote a downloaded subset into a complete report.
public struct PropertyManagementReportSnapshot: Encodable, Equatable, Sendable {
    public let reference: ProtectedArtifactSnapshotReference
    public let project: PropertyManagementReportProject
    public let provenance: PropertyManagementReportProvenance
    public let currency: CurrencyCode
    /// Includes every supplied parent and revision, even an unoccupied Space.
    public let spaces: [PropertyManagementReportSpace]
    public let groups: [PropertyManagementReportGroup]
    public let totals: PropertyManagementReportTotals
    public let sourceSetHash: ProtectedArtifactSHA256
    public let reportKind = "property_management"

    public static func build(project: PropertyManagementReportProject, spaces: [PropertyManagementReportSpace],
                             items: [PropertyManagementReportItem], currency: CurrencyCode,
                             provenance: PropertyManagementReportProvenance) throws -> Self {
        guard provenance.readiness == .ready else { throw PropertyManagementReportFailure.incompleteReadiness }
        guard project.accountId == provenance.accountId, project.projectId == provenance.projectId else {
            throw PropertyManagementReportFailure.scopeMismatch
        }
        guard project.revision > 0 else { throw PropertyManagementReportFailure.invalidRevision }
        var spaceIDs = Set<SpaceID>()
        for space in spaces {
            guard space.accountId == project.accountId, space.projectId == project.projectId else {
                throw PropertyManagementReportFailure.scopeMismatch
            }
            guard space.revision > 0 else { throw PropertyManagementReportFailure.invalidRevision }
            guard spaceIDs.insert(space.spaceId).inserted else { throw PropertyManagementReportFailure.duplicateSpace }
        }
        var itemIDs = Set<ItemID>()
        var placementIDs = Set<EntityID>()
        for item in items {
            guard item.accountId == project.accountId, item.projectId == project.projectId else {
                throw PropertyManagementReportFailure.scopeMismatch
            }
            guard item.itemRevision > 0 else { throw PropertyManagementReportFailure.invalidRevision }
            guard itemIDs.insert(item.itemId).inserted else { throw PropertyManagementReportFailure.duplicateItem }
            guard placementIDs.insert(item.placementId).inserted else { throw PropertyManagementReportFailure.duplicatePlacement }
            if let spaceId = item.spaceId, !spaceIDs.contains(spaceId) { throw PropertyManagementReportFailure.missingSpace }
            if item.accounting?.resolution == .accountedFor,
               let amount = item.marketValue, amount.currency != currency {
                throw PropertyManagementReportFailure.mixedCurrency
            }
            guard let accounting = item.accounting,
                  accounting.resolution != .relationshipEvidenceIncomplete else {
                throw PropertyManagementReportFailure.incompleteReadiness
            }
            guard accounting.evidence.accountId == item.accountId,
                  accounting.evidence.projectId == item.projectId,
                  accounting.evidence.itemId == item.itemId,
                  accounting.evidence.spaceId == item.spaceId else {
                throw PropertyManagementReportFailure.scopeMismatch
            }
        }
        // Locale-independent byte order, with exact identity as the tie-breaker.
        func ordered(_ nameA: String, _ idA: String, _ nameB: String, _ idB: String) -> Bool {
            let a = Array(nameA.utf8), b = Array(nameB.utf8)
            return a == b ? idA.utf8.lexicographicallyPrecedes(idB.utf8) : a.lexicographicallyPrecedes(b)
        }
        let sortedSpaces = spaces.sorted { ordered($0.name, $0.spaceId.rawValue, $1.name, $1.spaceId.rawValue) }
        let sortedItems = items.filter { $0.accounting?.resolution == .accountedFor }
            .sorted { ordered($0.name, $0.itemId.rawValue, $1.name, $1.itemId.rawValue) }
        func summarize(_ rows: [PropertyManagementReportItem]) throws -> PropertyManagementReportTotals {
            var known = Money.zero(currency: currency)
            var unknown = 0
            for row in rows {
                if let amount = row.marketValue { known = try known.adding(amount) } else { unknown += 1 }
            }
            return .init(itemCount: rows.count, knownMarketValueSubtotal: known, unknownMarketValueCount: unknown)
        }
        var groups: [PropertyManagementReportGroup] = []
        for space in sortedSpaces {
            let rows = sortedItems.filter { $0.spaceId == space.spaceId }
            if !rows.isEmpty { groups.append(.init(spaceId: space.spaceId, name: space.name, rows: rows, totals: try summarize(rows))) }
        }
        let unplaced = sortedItems.filter { $0.spaceId == nil }
        if !unplaced.isEmpty { groups.append(.init(spaceId: nil, name: "No Space", rows: unplaced, totals: try summarize(unplaced))) }
        let totals = try summarize(sortedItems)
        // Bind exclusion decisions too, without exporting excluded Item details
        // or their private accounting relationships in the report envelope.
        let accountingHash = try ProtectedArtifactSHA256.make(bytes: encode(items
            .sorted { $0.itemId.rawValue.utf8.lexicographicallyPrecedes($1.itemId.rawValue.utf8) }
            .map(\.accounting)))
        let sourceHash = try ProtectedArtifactSHA256.make(bytes: encode(SourceSet(project: project, spaces: sortedSpaces,
            items: sortedItems, accountingEvidenceHash: accountingHash)))
        let payload = Content(reportKind: "property_management", project: project, provenance: provenance, currency: currency,
            spaces: sortedSpaces, groups: groups, totals: totals, sourceSetHash: sourceHash)
        let hash = try ProtectedArtifactSHA256.make(bytes: encode(payload))
        let reference = try ProtectedArtifactSnapshotReference(snapshotID: .init(validating: String(hash.rawValue.prefix(32))),
            snapshotHash: hash, visibilityScopeID: provenance.visibilityScopeID,
            profileVersion: .init(validating: "property-management-v1"), authorityVersion: provenance.authorityVersion)
        return .init(reference: reference, project: project, provenance: provenance, currency: currency,
            spaces: sortedSpaces, groups: groups, totals: totals, sourceSetHash: sourceHash)
    }

    /// The canonical envelope includes `reference`. Its digest is NOT the
    /// snapshotHash; the reference binds the content bytes below without a
    /// circular self-hash.
    public func canonicalData() throws -> Data { try Self.encode(self) }

    /// Exact bytes hashed by reference.snapshotHash and used to derive its ID.
    public func canonicalContentData() throws -> Data {
        try Self.encode(Content(reportKind: reportKind, project: project, provenance: provenance, currency: currency,
            spaces: spaces, groups: groups, totals: totals, sourceSetHash: sourceSetHash))
    }

    private static func encode<T: Encodable>(_ value: T) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(value)
    }
    private struct SourceSet: Encodable {
        let project: PropertyManagementReportProject
        let spaces: [PropertyManagementReportSpace]
        let items: [PropertyManagementReportItem]
        let accountingEvidenceHash: ProtectedArtifactSHA256
    }
    private struct Content: Encodable {
        let reportKind: String
        let project: PropertyManagementReportProject
        let provenance: PropertyManagementReportProvenance
        let currency: CurrencyCode
        let spaces: [PropertyManagementReportSpace]
        let groups: [PropertyManagementReportGroup]
        let totals: PropertyManagementReportTotals
        let sourceSetHash: ProtectedArtifactSHA256
    }
}
